### PR TITLE
Moved the HydroPowerRecorder in to pywr.recorders.

### PR DESCRIPTION
**Pull request by jetuk**
This was some code in the hydrpower example. I have made it part of `pywr.recorders` so we can use it else where. Tests included.

### CODE REVIEW

**Comment by jetuk** [comment]
I have some doubts about this. I think it might be better as `Parameter` rather than recorder.

**Comment by jetuk** [comment]
Actually I've changed my mind. This should probably be two recorders,

1. Based on `NumpyArrayNodeRecorder` to keep the full timeseries, and
2. Based on `BaseConstantNodeRecorder` to keep the total value for a run.

**Comment by jetuk** [comment]
I've done what I said in the previous comment. There are now two recorders both cythonised:

- `HydroPowerRecorder(NumpyArrayNodeRecorder)`
- `TotalHydroEnergyRecorder(BaseConstantNodeRecorder)`

File: examples/hydropower_example.json [hunk 1]
```diff
@@ -27,10 +27,10 @@
             "max_volume": 200000,
             "initial_volume": 170000,
             "level":  {
-              "type": "interpolatedlevel",
+              "type": "interpolatedvolume",
               "node": "reservoir1",
               "volumes": [0, 25000, 50000, 75000, 100000, 150000, 200000],
-              "levels": [0, 29.2, 36.8, 42.2, 46.4, 53.1, 58.5],
+              "values": [0, 29.2, 36.8, 42.2, 46.4, 53.1, 58.5],
               "kind": "cubic"
```
Comment (Batch21) [inline examples/hydropower_example.json]: to get the example work I had to move the this parameter into the parameter section of the json.

File: examples/hydropower_example.json [hunk 2]
```diff
@@ -98,7 +98,7 @@
             "storage_node": "reservoir1",
             "level": 35.0,
```
Comment (Batch21) [inline examples/hydropower_example.json]: I had to change `level` to `turbine level` and replace `storage_node` with `water_elevation_parameter`, pointing to the interpolated volume parameter

File: pywr/recorders/_hydropower.pyx [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+                                             energy_unit_conversion=self.energy_unit_conversion)
+
+            self._values[scenario_index.global_id] += power * days
+
+    @classmethod
+    def load(cls, model, data):
+        from pywr.parameters import load_parameter
+        node = model._get_node_from_ref(model, data.pop("node"))
+        if "water_elevation_parameter" in data:
+            water_elevation_parameter = load_parameter(data.pop("water_elevation_parameter"))
```
Comment (Batch21) [inline pywr/recorders/_hydropower.pyx]: `model` needs to be the first argument to `load_parameter`

**Comment by jetuk** [comment]
Thanks @Batch21. I'll fix these tomorrow. The conflict is minor; to do with the imports in `test_recorders.py`

**Comment by jetuk** [comment]
I've addressed the issues. A simplified version of the example is now loaded and run as a test, but it has no values checked.

**Comment by Batch21** [comment]
The updates look good. I assume there is no need to check any values in the example test as this is covered in the other tests

**Comment by jetuk** [comment]
Good point. It should probably been done for completeness as that test as not assertions currently.

**Comment by jetuk** [comment]
I think this is done now. The commits can probably be squashed before merging. There's a slightly messy history with merging from master, etc.

**Comment by jetuk** [comment]
@Batch21 are you happy with this?

**Comment by Batch21** [comment]
yeah, all looks good

**Pull request merged**